proxyutil/proxy.go

```diff
@@ -0,0 +1,155 @@
+// Package proxyutil implements helper types and functions for proxying
+// (generally non-rpc) http requests to/from backing services.
+package proxyutil
+
+import (
+	"io"
+	"net/http"
+	"strings"
+)
+
+// The logic in this file is largely lifted from the implementation of
+// httputil.ReverseProxy
+
+// ReverseProxyClient is a wrapper around a basic *http.Client which performs
+// requests as a reverse http proxy instead of a normal client. This involves
+// adding/removing certain headers and dealing with Keep-Alive.
+type ReverseProxyClient struct {
+	// If nil then http.DefaultClient will beused
+	Client *http.Client
+}
+
+// Hop-by-hop headers. These are removed when sent to the backend.
+// http://www.w3.org/Protocols/rfc2616/rfc2616-sec13.html
+var hopHeaders = []string{
+	"Connection",
+	"Proxy-Connection", // non-standard but still sent by libcurl and rejected by e.g. google
+	"Keep-Alive",
+	"Proxy-Authenticate",
+	"Proxy-Authorization",
+	"Te",      // canonicalized version of "TE"
+	"Trailer", // not Trailers per URL above; http://www.rfc-editor.org/errata_search.php?eid=4522
+	"Transfer-Encoding",
+	"Upgrade",
+}
+
+func copyHeader(dst, src http.Header) {
+	for k, vv := range src {
+		for _, v := range vv {
+			dst.Add(k, v)
+		}
+	}
+}
+
+// Do will perform the given request, which should have been taken in from an
+// http.Handler and is now being forwarded on with a new URL set. If the
+// request's context is cancelled then the request will be cancelled.
+func (rp ReverseProxyClient) Do(req *http.Request) (*http.Response, error) {
+	cl := rp.Client
+	if cl == nil {
+		cl = http.DefaultClient
+	}
+
+	outreq := new(http.Request)
+	*outreq = *req // includes shallow copies of maps, but okay
+
+	outreq.Proto = "HTTP/1.1"
+	outreq.ProtoMajor = 1
+	outreq.ProtoMinor = 1
+	outreq.Close = false
+
+	// Remove hop-by-hop headers to the backend. Especially
+	// important is "Connection" because we want a persistent
+	// connection, regardless of what the client sent to us. This
+	// is modifying the same underlying map from req (shallow
+	// copied above) so we only copy it if necessary.
+	copiedHeaders := false
+	for _, h := range hopHeaders {
+		if outreq.Header.Get(h) != "" {
+			if !copiedHeaders {
+				outreq.Header = make(http.Header)
+				copyHeader(outreq.Header, req.Header)
+				copiedHeaders = true
+			}
+			outreq.Header.Del(h)
+		}
+	}
+
+	AddXForwardedFor(outreq, req)
+
+	res, err := cl.Do(outreq)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, h := range hopHeaders {
```

Comment:
It's deleting the headers off the response, not the request. It's what httputil did, I didn't really ask questions >_>